File: src/content/docs/topics/cs/design-patterns/decorator.go
package main

import "fmt"

type DataService interface {
	FetchData(key string) string
}

// Compile-time interface checks.
var _ DataService = (*LoggerDecorator)(nil)
var _ DataService = (*CacheDecorator)(nil)

type RealDataService struct{}

func (r *RealDataService) FetchData(key string) string {
	fmt.Printf("  [DB] Fetching key: %s\n", key)
	return "value_for_" + key
}

type LoggerDecorator struct {
	wrapped DataService
}

func (l *LoggerDecorator) FetchData(key string) string {
	fmt.Printf("[LOG] FetchData called with key=%q\n", key)
	result := l.wrapped.FetchData(key)
	fmt.Printf("[LOG] FetchData returned %q\n", result)
	return result
}

type CacheDecorator struct {
	wrapped DataService
	cache   map[string]string
}

func NewCacheDecorator(wrapped DataService) *CacheDecorator {
	return &CacheDecorator{wrapped: wrapped, cache: make(map[string]string)}
}

func (c *CacheDecorator) FetchData(key string) string {
	if val, ok := c.cache[key]; ok {
		fmt.Printf("[CACHE] hit for %q\n", key)
		return val
	}
	result := c.wrapped.FetchData(key)
	c.cache[key] = result
	fmt.Printf("[CACHE] stored %q\n", key)
	return result
}

func main() {
	var service DataService = NewCacheDecorator(
		&LoggerDecorator{wrapped: &RealDataService{}},
	)

	fmt.Println("--- First call ---")
	service.FetchData("user:42")

	fmt.Println("--- Second call ---")
	service.FetchData("user:42")
}

File: src/content/docs/topics/cs/design-patterns/decorator.py
from __future__ import annotations
from abc import ABC, abstractmethod


class DataService(ABC):
    @abstractmethod
    def fetch_data(self, key: str) -> str: ...


class RealDataService(DataService):
    def fetch_data(self, key: str) -> str:
        print(f'  [DB] Fetching key: {key}')
        return f'value_for_{key}'


class LoggerDecorator(DataService):
    def __init__(self, wrapped: DataService) -> None:
        self._wrapped = wrapped

    def fetch_data(self, key: str) -> str:
        print(f'[LOG] fetch_data called with key="{key}"')
        result = self._wrapped.fetch_data(key)
        print(f'[LOG] fetch_data returned "{result}"')
        return result


class CacheDecorator(DataService):
    def __init__(self, wrapped: DataService) -> None:
        self._wrapped = wrapped
        self._cache: dict[str, str] = {}

    def fetch_data(self, key: str) -> str:
        if key in self._cache:
            print(f'[CACHE] hit for "{key}"')
            return self._cache[key]
        result = self._wrapped.fetch_data(key)
        self._cache[key] = result
        print(f'[CACHE] stored "{key}"')
        return result


service: DataService = CacheDecorator(LoggerDecorator(RealDataService()))

print('--- First call ---')
service.fetch_data('user:42')

print('--- Second call ---')
service.fetch_data('user:42')
